### DISCUSSION TITLE
[New Component Request] Diagramming component set (componentize Automations?)

### DISCUSSION THREAD

**Discussion by chrisshaw** [Budibase/budibase#2742] [Ideas]
> I know I'm probably going to be waiting a while for this but... :pray: :pray: :pray: :pray: :pray: :pray: 

Have you considered "componentizing" the Automations components themselves? Or providing a set of diagram-building components?

Literally any diagramming components would suffice, but I also thought that it may be easier to grab your existing Automations components and expose them to the Budibase designer.

## Why?

For me personally, I have 2 use cases:

- **Drawing dependencies between different rows:** We have an entity called Opportunities that our agents can pursue, but some opportunities **depend** on the status of other opportunities. The Automation's workflow would allow some of our less tech-savvy team members a visual way to string these different rows together and provide a faster way to string multiple rows together all at once, rather than going into a record detail page updating a field on each object.
- **Agent scripts:** We have entities called Messages and Responses that are used as helpers for common scenarios chat agents fine themselves in. Our agents find it's often useful to have these all together in one full process flow. That way, they can (a) use the helper responses as canned responses and (b) know where their conversation _should_ go, treating the process flow as an example. (This is, of course, just an instance of the above use case, but I figure I'd call it out separately because it places heavier emphasis on using these components from a Read/View perspective rather than Edit.)

But these components could fill the purpose for most diagramming, process, or workflow authoring use cases, which to me feel right at home in an internal tool.

**Comment by mjashanks** [comment]
Thanks Chris. 

It feels like you are asking for simple image of a process flow! But I assume there is more to it than that. 

Should there be some sort of interactivity in each "step". What would users be able to do with each step? Are they forms?

**Reply by chrisshaw** [reply]
Yes, you're right, there would be interactivity for each step. And yes, they're essentially forms. (I should mention that, while I haven't looked into it, but I would bet both of these are possible today in BB; just not very convenient.)

There's really multiple uses here that have different kinds of functionality.

### Workflow builder

Here I'm basically just saying, "Can you give us a Zapier-like interface to specify workflows?" This is (I believe) really just like saying, "Can you give us forms that to let us fill in relationship fields through a node and edge representation?"

For example, let's say an Opportunity represents a conversation a chat agent can have with a user. Let's say that:

```ts
interface Opportunity {;
  status: 'open' | 'closed';
  dependsOn: Opportunity;
}
```

Using the new components in our workflow builder, we could create two new Opportunities at once, set the first one to `open`, and then, **via an edge,** set the `dependsOn` value to the second Opportunity.

The workflows themselves would be entirely done on the backend (just like Zapier). We would likely write Automations based on record changes. If the agent changes that Opportunity's `status` field to, say, `closed`, it would trigger an Automation to look at the Opportunity in the `dependsOn` field and set it to `open`. So the workflow builder is really just a form where we are modifying multiple rows at once, and we're able to update relationship fields via a more visual UX.

### Agent script

This one's more involved (I think). Here I'm saying, "Could you replace [Salesforce](https://www.salesforce.com/products/platform/solutions/automate-business-processes/) [Flow](https://trailhead.salesforce.com/en/content/learn/trails/automate_business_processes) for me?"

Salesforce Flow can do a lot (although it's incredibly tedious to build with). Today in Flow, we have Scripts that look like this.

![image](https://user-images.githubusercontent.com/2966918/134963830-a9072be1-9768-4042-b9c5-abdd89a4d113.png)

![image](https://user-images.githubusercontent.com/2966918/134964332-43f93b21-f9c5-42bd-b4aa-a64828c62773.png)

What you'll see in there:

- Forms
- Logic branches
- CRUD actions
- Setting contained "Flow" state (like app state, but for this specific run of a Flow)
- Calling other Flows
- (Not pictured) Calling external endpoints

So it's a bit like turning Budibase inside out, where you focus on the Actions and Conditions first and then embed the definition of the UI.

**For my use case as an agent script, though, I think we could abandon these complex flows and get away with a dynamically-rendered form that sends outbound requests.**

Let's say we have:

```ts
interface ScriptInput {
  kind: 'text' | 'choice'
  value: string;
  defaultValue: string;
}

interface ScriptComponent {
  question: string;
  input: ScriptInput
  priority: number;
}

interface Script {
  inputs: ScriptComponent[]
}
```

We still have a builder experience, where we need to _write_ the script. That's equivalent to stacking ScriptComponents.

What makes this a bit more interesting, though, is that our scripts aren't really "scripts" but checklists. We don't expect agents to stay on script; it just helps them understand what data to collect or points to make by the end of a conversation. It's often helpful, then, for agents to have the full context up front as a flowchart, so they can get an idea of how an ideal conversation should go.

#### Oh, things just got interesting :point_down: 

I've left out the concept of conditions and actions because it got complex. But those would be a nice to have for the workflow builder and a high priority for the script builder. That's perhaps the most interesting quality of this feature request: the ability to store actions and conditions in our own tables and bind them to the UI like any other value.

**Reply by mjashanks** [reply]
I get you. I have a background in healthcare software, whereby healthcare professionals were doing triage, with patients, over the phone.

They had a similar scripted experience, but couldn't be too rigid as they wanted to have a pleasant relationship with the patient, and let the conversation go in different directions. They were also well trained medical professionals and didn't like to be restricted.

They also had some "Red Flag Answers", whereby the flow had to be directed to certain high priority sets of questions.

One thing that we have recently thought about creating in Budibase is the concept of "Templated Sections", which would make it much easier for you to place sets of components on the screen. That could make the job of putting this together much quicker.

It's fantastic to get this level of detail from you - thank you. I don't know where we will go with this, but this is great background to have.

**Reply by chrisshaw** [reply]
Yes, exactly. We're dealing with minors so we have similar requirements to healthcare and our agents.

My first impression of templated sections is of custom components that we build from pre-built components, which would be nice. One issue I still see is that we have script writers: users who don't need to be app developers with full access but still need to be able to build parts of the app with these templated sections to make scripts. So I believe we'd need to pair the new templated sections with an updated access model that allows "partial developers."
